taichi/backends/codegen_metal.cpp

```diff
@@ -410,6 +388,59 @@ class MetalKernelCodegen : public IRVisitor {
     emit("  static_assert(sizeof(T) == sizeof(G), \"Size mismatch\");");
     emit("  return *reinterpret_cast<thread const T*>(&g);");
     emit("}}");
+    emit("");
+
+    gen_ifloordiv();
+    gen_fatomic_fetch_add_func();
+  }
+
+  void gen_ifloordiv() {
+    constexpr char kLhs[] = "lhs";
+    constexpr char kRhs[] = "rhs";
+    emit("inline int {}(int {}, int {}) {{", kIFloordivFunc, kLhs, kRhs);
+    push_indent();
+    constexpr char kIntm[] = "intermediate";
+    emit("const int {} = ({} / {});", kIntm, kLhs, kRhs);
+    // Should we construct an AST for this?
+    const auto expr_str = fmt::format(
+        "(({lhs} * {rhs} < 0) && ({rhs} * {intm} != {lhs})) ? ({intm} - 1) "
+        ": {intm}",
+        fmt::arg("lhs", kLhs), fmt::arg("rhs", kRhs), fmt::arg("intm", kIntm));
+    emit("return ({});", expr_str);
+    pop_indent();
+    emit("}}");
+    emit("");
+  }
+
+  void gen_fatomic_fetch_add_func() {
```

Comment:
This does look cleaner than the original approach!

Will that make things even cleaner, if we directly emit a multiline string of the function `fatomic_fetch_add`, or even put all these functions into a "header" and then "include" it? I.e. directly include the emitted shader code by `gen_fatomic_fetch_add_func` instead of emitting and formatting the lines.

I've got good experience when taking this approach in the legacy backend. E.g. here I implement CPU atomics using CAS
https://github.com/taichi-dev/taichi/blob/1cc91f52acebb2bb0e39a49baecbb459b60389d7/taichi/arithmetics.h#L1045 and this file is then directly included as a header. Not sure if this is a good idea on the Metal backend though.

Reply:
Thanks for the suggestion! While I don't like to emit these function as string literals either, unfortunately, I don't think that can work on Metal, because of the usage of [`newLibraryWithSource:options:completionHandler`](https://developer.apple.com/documentation/metal/mtldevice/1433351-newlibrarywithsource). The doc says that 

> The source may only import the Metal standard library. There is no search path to find other functions.

X(

Reply:
Let me remove all those `fmt::format()` stuffs and just emit a multiline string lietral..

Reply:
To me, it's now much easier to read! It's a pity that Metal shaders cannot do `#include`... If we are getting more of these functions, we might consider manually "including" something like `helpers.metal` by prepending to the emitted code, so that we get rid of the `emit("...")`. Of course, you know the Metal backend much better than I do so it's your decision. I'm just talking :-)

Reply:
Great :-)

> Metal shaders cannot do #include... If we are getting more of these functions

Ah, Metal shaders can, but then you will have to output these source code into an actual file, then invoke the Metal compiler, which I think is the legacy approach? We can switch to that if these helpers continue to grow.

> we might consider manually "including" something like helpers.metal by prepending to the emitted code

+1. I guess one way to do this is the `inc.h` approach, where we put these helpers into a header, surrounded by some macro. And during the codegen time, that macro expands to `emit()`... Will need more time playing on this idea...

Reply:
> Will need more time playing on this idea...

Yeah I thought a little bit about this. One solution could be creating a `helpers.cpp` with contents like

```C++
std::string metal_helpers = R("

f32 atomic_add(f32 *dest, f32 val) {
  do (...) {
    ...  
  } while (CAS...);
}

...other functions...
");
```

then we prepend `metal_helpers` to the emitted code. This is very close to invoke a single `emiit` with a multiline string literal though. Anyway, I think we are good for now. Merging...

Reply:
One sec, let me obey the new rules and squash it 😂

Reply:
(Sorry about merging too fast... XD)

> Yeah, i think the macro can expand to std::string metal_helpers = R(" only inside codegen, and will expand to nothing in other cases.

That's a good idea! I think having a *header* with just the function definition as the body (without the quotes) will enable the IDE to syntax highlight to some extent.

>  Then we can even use c++ compiler to make sure the code compilers.. (On the other hand, there are a few Metal specific keywords and functions, so maybe this approach will be very noisy)

I agree having the C++ compile checking syntax correctness for us is good, but as you said it might be noisy...

Reply:
Btw it seems that there's no way for me to merge in https://github.com/taichi-dev/taichi/pull/465/commits/a753899b29b25fbc157e1562af7497aa798e9b1f ... Please feel free to open a new PR for that (so sorry about the trouble...)

Reply:
I think that commit belongs to my local branch, and on master these two commits are already squashed into one? See abbf5b13537ab5e0f1d951fb3502b65a4f509bb1 :)